File: sapphire/examples/hanksTodoRuby/src/main/ruby/sapphire/appexamples/hankstodo/todo_list_manager.rb
#!/usr/bin/ruby
# TODO: Support multifile SO .
# Ruby library/file load steps:
#   loadPath = File.expand_path('src/main/ruby/sapphire/appexamples/hankstodo',__dir__)
#   $:.unshift(folder) unless $:.include?(loadPath)
# Identified Issue: When Ruby SO is used in kernel server "__dir__" returns current working directory
#                   But when run in Ruby env it returns directory of ruby file itself.
class TodoList
    def initialize(name)
        @to_dos = Array.new
        @name = name
    end

    def addTodo(todo)
        @to_dos.push(todo)
        puts "add #{todo} success!!!"
        return "OK!"
    end

    def completeTodo(todo)
        @to_dos.delete(todo)
        puts "#{todo} task removed"
        return "OK!"
    end

    def getTodos
        return @to_dos
    end

    def getName
        @name
    end
end

class Manager
    def initialize
        @todo_lists = Array.new
    end

    def getList(name)
        return @todo_lists[name]
    end
end

class TodoListManager < Manager
    def newTodoList(name)
        list = nil
        @todo_lists.each { |lst|
            if lst.getName == name
                list = lst
            end
        }

        if list == nil
            list = TodoList.new(name)
            @todo_lists << list
        end

        puts "Create list"
        return name
    end

    def deleteTodoList(name)
        list = nil
        @todo_lists.each { |lst|
            if lst.getName == name
                list = lst
            end
        }

        if list != nil
            @todo_lists.delete(list)
        end

        puts "List Deleted"
        return "OK!!!!"
    end

    def addTodo(name, todo)
        list = nil
        @todo_lists.each { |lst|
        puts "list name #{lst.getName}"
            if lst.getName == name
                puts "find list success !!!"
                list = lst
            end
        }

        if list != nil
            return list.addTodo(todo)
        end
        return "Failed !!!!"
    end

    def getTodos(name)
        list = nil
        @todo_lists.each { |lst|
            if lst.getName == name
                puts "find list success !!!"
                list = lst
            end
        }

        if list != nil
            return list.getTodos
        end

        return nil
    end

    def testSpacialCharInFunctionName?(str)
        return str;
    end

    def completeTodo(name, todo)
        list = nil
        @todo_lists.each { |lst|
            if lst.getName == name
                puts "find list success !!!"
                list = lst
            end
        }

        if list != nil
            return list.completeTodo(todo)
        end

        return nil
    end
end

File: sapphire/examples/hanksTodoRuby/src/main/ruby/sapphire/appexamples/hankstodo/todo_list_manager.js
class TodoListManager {
    constructor() {
        console.log("constructor");
        this.map = new Map();
    }

    new_todo_list(name) {
        console.log("new_todo_list");
        var b = this.map.has(name);
        if (!b) this.map.set(name, new Set());
        return b;
    }

    delete_todo_list(name) {
        console.log("delete_todo_list");
        var b = this.map.has(name);
        if (b) this.map.delete(name);
        return b;
    }

    add_todo(todoListName, todoTaskName) {
        console.log("add_todo");
        this.map.get(todoListName).add(todoTaskName);
        return "OK_add_todo"
    }

    get_todos(todoListName) {
        console.log("get_todos");
        var v = this.map.get(todoListName);
        var res = [];
        v.forEach(function(value) {
            res.push(value);
        });
        return res;
    }

    complete_todo(todoListName, todoTaskName) {
        console.log("complete_todo");

        if (this.map.has(todoListName) && this.map.get(todoListName).has(todoTaskName)) {
            this.map.get(todoListName).delete(todoTaskName);
            return true;
        }

        return false;
    }
}

function test() {
    var c = new TodoListManager();
    c.new_todo_list("todoList1");
    c.add_todo("todoList1", "task1");
    c.add_todo("todoList1", "task2");
    var v = c.get_todos("todoList1");
    v.forEach(function(value) {
      console.log(value);
    });
    c.complete_todo("todoList1", "task1");
    v = c.get_todos("todoList1");
    v.forEach(function(value) {
          console.log(value);
        });
    c.delete_todo_list("todoList1");
}

test();